src/lib/support/StringCircularBuffer.h

```diff
@@ -0,0 +1,76 @@
+/*
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <cstddef>
+
+#include <core/CHIPError.h>
+#include <support/Span.h>
+
+namespace chip {
+
+/**
+ * @brief A circular buffer to store strings.
+ *
+ * This circular buffer provides a queue like interface push/pop strings from
+ * the buffer. When there is no space to push a new string, the oldest string
+ * will be discarded silently to free up space.
+ *
+ * NOTE: this class never handles trailing zero.
+ */
+class StringCircularBuffer
+{
+public:
+    /**
+     * @brief Create a string circular buffer.
+     *
+     * @param storage  The underlying storage. This class doesn't take the ownership of the storage.
+     * @param capacity The length of the storage.
+     */
+    StringCircularBuffer(uint8_t * storage, size_t capacity) : mStorage(storage), mCapacity(capacity), mDataStart(0), mDataEnd(0) {}
+
+    CHIP_ERROR Push(const ByteSpan & payload);
+    CHIP_ERROR Pop();
+
+    bool IsEmpty() const;
+
+    /** @brief get the length of the front element. */
+    size_t GetFrontSize() const;
+
+    /** @brief read the front element into dest. The length of dest must be equal or larger than GetFrontSize */
+    CHIP_ERROR ReadFront(uint8_t * dest) const;
```

Comment:
How about giving a ByteSpan as destination? Then we can do error checking and not have these "you pass buffer and ensure size is correct"

Reply:
> How about giving a ByteSpan as destination? 

At the moment ByteSpan only holds `const uint8_t*`....

Reply:
That makes me sad :( I can see a need of wanting both const and modifiable spans.

Reply:
Yes, I think we should make it possible to write to a span's buffer...  Current spans were added in places where we were dealing with read-only buffers, so they catered to that use case, but we've had a few places come up where writable buffers plus lengths are being passed around.

Reply:
If the `Span` is mimicking the behavior of `std::span`, then it shouldn't be const. From C++ spec:

```
constexpr pointer data() const noexcept;
```

`constexpr` is qualifying the pointer itself, not the pointer type.

Reply:
PR #7642 to address this